jina/orchestrate/flow/base.py

```diff
@@ -2368,24 +2366,97 @@ def block(
         :param stop_event: a threading event or a multiprocessing event that onces set will resume the control Flow
             to main thread.
         """
+
+        def _restart_flow(changed_file):
+            self.logger.info(
+                f'change in Flow YAML {changed_file} observed, restarting Flow'
+            )
+            self.__exit__(None, None, None)
+            new_flow = Flow.load_config(changed_file)
+            self.__dict__ = new_flow.__dict__
+            self.__enter__()
+
+        def _restart_deployment(deployment, changed_file):
+            self.logger.info(
+                f'change in Executor configuration YAML {changed_file} observed, restarting Executor deployment'
+            )
+            deployment.__exit__(None, None, None)
+            old_args, old_needs = deployment.args, deployment.needs
+            new_deployment = Deployment(old_args, old_needs)
+            deployment.__dict__ = new_deployment.__dict__
+            deployment.__enter__()
+
         try:
-            if stop_event is None:
-                self._stop_event = (
-                    threading.Event()
-                )  #: this allows `.close` to close the Flow from another thread/proc
-                if not __windows__:
-                    self._stop_event.wait()
-                else:
+            watch_changes = self.args.restart or any(
+                [
+                    deployment.args.restart
+                    for deployment in list(self._deployment_nodes.values())
+                ]
+            )
+            watch_files_from_deployments = {}
+            for name, deployment in self._deployment_nodes.items():
+                if deployment.args.restart:
+                    if deployment._is_executor_from_yaml:
+                        watch_files_from_deployments[deployment.args.uses] = name
+            watch_files_list = list(watch_files_from_deployments.keys())
+
+            config_loaded = getattr(self, '_config_loaded', '')
+            if config_loaded.endswith('yml') or config_loaded.endswith('yaml'):
+                watch_files_list.append(config_loaded)
+
+            if watch_changes and len(watch_files_list) > 0:
+
+                with ImportExtensions(
+                    required=True,
+                    logger=self.logger,
+                    help_text='''restart requires watchfiles dependency to be installed. You can do `pip install 
+                    watchfiles''',
+                ):
+                    from watchfiles import watch
+
+                new_stop_event = stop_event or threading.Event()
+                if len(watch_files_list) > 0:
                     while True:
```

Comment:
we should remove the while, for changes in watch should take care of waiting

Reply:
it is sync, after the first change is gone, it leaves the for loop I believe

Reply:
at least it was my observation, the test does not pass in the second change

Reply:
Now I checked and removed